ee/performance/stats.go

```diff
@@ -56,51 +62,121 @@ func ProcessStatsForPid(ctx context.Context, pid int) (*PerformanceStats, error)
 		return nil, fmt.Errorf("getting process handle for pid %d: %w", pid, err)
 	}
 
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+
+	ps, memInfo, err := statsForProcess(ctx, proc)
+	if err != nil {
+		return nil, fmt.Errorf("gathering stats for process: %w", err)
+	}
+
+	ps.MemInfo.NonGoMemUsage = nonGoMemUsage(memInfo, &memStats)
+	ps.MemInfo.GoMemUsage = goMemUsage(&memStats)
+	ps.MemInfo.HeapTotal = heapTotal(&memStats)
+
+	// Record stats
+	observability.GoMemoryUsageGauge.Record(ctx, int64(ps.MemInfo.GoMemUsage))
+	observability.NonGoMemoryUsageGauge.Record(ctx, int64(ps.MemInfo.NonGoMemUsage))
+	observability.MemoryPercentGauge.Record(ctx, int64(ps.MemInfo.MemPercent))
+	observability.CpuPercentGauge.Record(ctx, int64(ps.CPUPercent))
+	observability.RSSHistogram.Record(ctx, int64(ps.MemInfo.RSS))
+
+	return ps, nil
+}
+
+// CurrentProcessChildStats gets memory and CPU stats for the current process's child processes;
+// it has the side effect of recording those metrics. (Pretty much any time we
+// do the work of collecting these stats, we want to emit measurements to our
+// metrics instruments.)
+func CurrentProcessChildStats(ctx context.Context) ([]*PerformanceStats, error) {
+	pid := os.Getpid()
+	return ChildProcessStatsForPid(ctx, int32(pid))
+}
+
+func ChildProcessStatsForPid(ctx context.Context, pid int32) ([]*PerformanceStats, error) {
+	proc, err := process.NewProcessWithContext(ctx, pid)
+	if err != nil {
+		return nil, fmt.Errorf("getting process handle for pid %d: %w", pid, err)
+	}
+
+	childProcesses, err := proc.ChildrenWithContext(ctx)
+	// ChildrenWithContext uses pgrep, which will exit with exit status 1 if there were no matching processes
+	// (i.e. no child processes). This is unexpected for us -- launcher should typically have child processes --
+	// but it's not necessarily an error. Only return an error if we got an actual error string back here;
+	// callers can handle an empty list of children appropriately in the case that we received exit status 1.
+	if err != nil && !strings.Contains(err.Error(), "exit status 1") {
+		return nil, fmt.Errorf("getting child processes for pid %d: %w", pid, err)
+	}
+
+	stats := make([]*PerformanceStats, 0)
+	for _, childProcess := range childProcesses {
+		ps, _, err := statsForProcess(ctx, childProcess)
+		if err != nil {
+			continue
+		}
+		stats = append(stats, ps)
+
+		if strings.Contains(ps.Cmdline, "osquery") {
+			observability.OsqueryRssHistogram.Record(ctx, int64(ps.MemInfo.RSS))
+			observability.OsqueryCpuPercentHistogram.Record(ctx, ps.CPUPercent)
+		}
+
+		// We want to grab one more level of child processes, to account for the desktop process
```

Comment:
just verifying- are we sure the CPU usage is not bubbled up/accounted for by the parent process stats? like if we have launcher proper -> launcher desktop sudo wrapper -> launcher desktop, those will all report separate CPU usages that we can sum together later?

Reply:
I think yes -- I consistently see the `sudo --preserve-env -u <username> /path/to/launcher desktop` process reporting a smaller CPU percent than its child desktop process. e.g. in a recent log, the parent desktop process had cpu_percent 0.017492090009365427 while the child had 0.39365630681368013. I can document this.

Reply:
Added a comment!